elf_load.c

```diff
@@ -311,13 +328,228 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+/* If vastart + phdr.p_filesz (vastart) < vastart + phdr.p_memsz (vaend),
+ * 0 out that remainder, either through memset or through anonymous mappings
+ */
+static int elf_load_mmap_filesz_memsz_diff(struct elf_prog *elf_prog,
+					   GElf_Phdr *phdr __maybe_unused,
+					   uintptr_t vastart, uintptr_t vaend)
+{
+	UK_ASSERT(elf_prog && phdr && vastart && vaend);
+
+	uk_pr_debug("%s: Zeroing 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)(vastart),
+		    (uint64_t)(vaend));
+
+	/* From the last byte contained in filesz to the last
+	 * byte contained in memsz, we can either have:
+	 * 1. a page alignment adjustment, where the end
+	 * address is just PAGE_ALIGN_UP(paddr + filesz), which
+	 * means that our initial call to mmap already read in
+	 * a page for us whose remaining bytes we memset without
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  -1, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
+				GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart_old, vastart_aligned;
+	uintptr_t vaend_old, vaend_new;
+	uintptr_t vastart, vaend;
+	__sz mmap_len;
+	int rc;
+
+	/* First start/vabase can't be !0 before loading first PT_LOAD */
+	UK_ASSERT(!elf_prog->start && !elf_prog->vabase && phdr);
+	/* PT_LOAD p_vaddr/p_paddr must be multiple of page size */
+	UK_ASSERT(!(phdr->p_vaddr & (phdr->p_align - 1)));
+	UK_ASSERT(!(phdr->p_paddr & (phdr->p_align - 1)));
+
+	mmap_len = prog_upperl + prog_lowerl + pt_load_max_align;
```

Comment:
Should this be `prog_upperl - prog_lowerl + pt_load_max_align`? Anyway it doesn't break anything since `prog_lowerl` is always 0.

Reply:
oh yes, it should be, sorry, I wrote this in a rush :man_facepalming:. I actually was unfortunate enough to find `ET_DYN` whose image base was not 0... Thank you! :love_you_gesture:

Reply:
Oh, wait, actually no, I just remembered. Let me show you this binary with an image base != 0 that I was talking about:
```
Elf file type is DYN (Position-Independent Executable file)
Entry point 0x473c40
There are 11 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x0000000000000268 0x0000000000000268  R      0x1000
  INTERP         0x0000000000000fe4 0x0000000000400fe4 0x0000000000400fe4
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  NOTE           0x0000000000000f80 0x0000000000400f80 0x0000000000400f80
                 0x0000000000000064 0x0000000000000064  R      0x4
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000002008870 0x0000000002008870  R E    0x1000
  LOAD           0x0000000002009000 0x0000000002409000 0x0000000002409000
                 0x000000000262b7e8 0x000000000262b7e8  R      0x1000
  LOAD           0x0000000004635000 0x0000000004a35000 0x0000000004a35000
                 0x0000000001bc4cd0 0x0000000001bc4cd0  RW     0x1000
  GNU_RELRO      0x0000000004635000 0x0000000004a35000 0x0000000004a35000
                 0x0000000001bc4cd0 0x0000000001bc4cd0         0x1000
  LOAD           0x00000000061fa000 0x00000000065fa000 0x00000000065fa000
                 0x00000000001383e0 0x00000000001b2970  RW     0x1000
  DYNAMIC        0x00000000061fed20 0x00000000065fed20 0x00000000065fed20
                 0x00000000000000c0 0x00000000000000c0  RW     0x8
  TLS            0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000008  R      0x8
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x8

 Section to Segment mapping:
  Segment Sections...
   00     
   01     .interp 
   02     .note.go.buildid 
   03     .text .plt .interp .note.go.buildid 
   04     .rodata .rela .rela.plt .gnu.version_r .dynstr .gnu.version .hash .dynsym 
   05     .data.rel.ro .data.rel.ro.typelink .data.rel.ro.itablink .data.rel.ro.gosymtab .data.rel.ro.gopclntab 
   06     .data.rel.ro .data.rel.ro.typelink .data.rel.ro.itablink .data.rel.ro.gosymtab .data.rel.ro.gopclntab 
   07     .go.buildinfo .got.plt .dynamic .got .noptrdata .data .bss .noptrbss 
   08     .dynamic 
   09     .tbss 
   10     
   ```
   
   Now it is useful to also add `prog_lowerl` because you can fool the interpreter afterwards, because it will add that `prog_lowerl` to the actual `vabase` :smile:. What do you think? :thinking:

Reply:
> Now it is useful to also add prog_lowerl because you can fool the interpreter afterwards, because it will add that prog_lowerl to the actual vabase 😄. What do you think? 🤔

Thanks, I understand it now. But should it really be `prog_upperl + prog_lowerl`?

I think `prog_upperl` is enough, because with `prog_upperl`, we are assuming that the `prog_lowerl` is zero, and we can directly get the `vabase`.

Reply:
Ah, you're right, apologies for my mistake. Can't understand how I misinterpreted this calculation :laughing:. I had to draw a diagram just to make sure I am not tripping.
![image](https://github.com/unikraft/app-elfloader/assets/45634497/68e7c5fc-a2f0-49df-b843-db505d7ce1d4)

Yep, `prog_upperl` should suffice :+1: Hah :triumph:.

Reply:
Done, thanks a lot! :100: